### PR TITLE
Added a boolean in msal-node's system options, which disables built-in retry policies

### LINKED ISSUE #6901: Add an option to disable MSAL Node's internal retry policy.
### Core Library

MSAL Node (@azure/msal-node)

### Wrapper Library

Not Applicable

### Public or Confidential Client?

Confidential

### Description

Add an option to disable MSAL Node's internal retry policy.

```json
 system: {
             disableInternalRetries: <boolean> // default FALSE
}
```

And update the docs to state: 

"Disables the MSAL built-in retry policies, allowing the app developer to specify their own retry policy. Currently only managed identity flows have a simple retry policy of retrying once on certain HTTP error codes."

### Source

Internal (Microsoft)

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**

### CODE REVIEW

File: lib/msal-node/docs/configuration.md [hunk 1]
```diff
@@ -94,12 +94,13 @@ const msalInstance = new PublicClientApplication(msalConfig);
...
+| `customAgentOptions`     | Set of configurable options to set on a http(s) agent                                                                                                                                                                                                | Object - [NodeJS documentation on alloweable options](https://nodejs.org/docs/latest-v16.x/api/http.html#new-agentoptions) | Empty Object `{}`                                                                                                                    |
+| `disableInternalRetries` | A boolean that, if enabled, disables MSALJS's built-in retry policies, allowing the app developer to specify their own retry policy. Currently, only Managed Identity flows have a simple retry policy of retrying once on certain HTTP error codes. | boolean                                                                                                                    | boolean `false`                                                                                                                      |
```
Comment (Robbie-Microsoft) [inline lib/msal-node/docs/configuration.md]: only the "disableInternalRetries" line was changed here

Comment (bgavrilMS) [inline lib/msal-node/docs/configuration.md]:
```suggestion
| `disableInternalRetries` | A flag that disables MSALJS's built-in retry policies, allowing the app developer to specify their own retry policy. Currently, only Managed Identity flows have a retry policy. | boolean                                                                                                                    | boolean `false`                                                                                                                      |
```

**Review by bgavrilMS — APPROVED** [review]

File: lib/msal-node/src/client/ManagedIdentityApplication.ts [hunk 2]
```diff
@@ -134,6 +134,7 @@ export class ManagedIdentityApplication {
         }
 
         const managedIdentityRequest: ManagedIdentityRequest = {
+            disableInternalRetries: this.config.system.disableInternalRetries,
```
Comment (neha-bhargava) [inline lib/msal-node/src/client/ManagedIdentityApplication.ts]: Should this variable be defined per application and not per request? @bgavrilMS

File: lib/msal-node/test/client/ManagedIdentitySources/Imds.spec.ts [hunk 3]
```diff
@@ -452,6 +452,41 @@ describe("Acquires a token successfully via an IMDS Managed Identity", () => {
 
                 jest.restoreAllMocks();
             });
+
+            test("ensures that a retry does not happen when the http status code from a failed network response is included in the retry policy, but the retry policy has been disabled", async () => {
```
Comment (neha-bhargava) [inline lib/msal-node/test/client/ManagedIdentitySources/Imds.spec.ts:456]: Nice :)

File: lib/msal-node/src/client/ManagedIdentityApplication.ts (hunk 2 above)
Comment (neha-bhargava) [reply lib/msal-node/src/client/ManagedIdentityApplication.ts]: I think this variable should be defined for the ManagedIdentityApplication and not ManagedIdentityRequest.

File: lib/msal-node/src/request/ManagedIdentityRequestParams.ts [hunk 4]
```diff
@@ -9,6 +9,7 @@
  * - resource  - resource requested to access the protected API. It should be of the form "{ResourceIdUri}" or {ResourceIdUri/.default}. For instance https://management.azure.net or, for Microsoft Graph, https://graph.microsoft.com/.default
  */
 export type ManagedIdentityRequestParams = {
+    disableInternalRetries?: boolean;
```
Comment (bgavrilMS) [inline lib/msal-node/src/request/ManagedIdentityRequestParams.ts]: Isn't this part of the public API? It should not be. 

We should not allow app dev to make the same config twice.

File: lib/msal-node/src/client/ManagedIdentityApplication.ts (hunk 2 above)
Comment (bgavrilMS) [reply lib/msal-node/src/client/ManagedIdentityApplication.ts]: `this.config.system.disableInternalRetries` should be the only config out there.

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts [hunk 5]
```diff
@@ -150,20 +152,24 @@ export abstract class BaseManagedIdentitySource {
                 networkRequest.computeParametersBodyString();
         }
 
+        const networkClient = managedIdentityRequest.disableInternalRetries
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts]: Isn't there an easier way to inject the `disableInternalRetries` in this class ? Similar to how you inject networkClient?

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts [hunk 6]
```diff
@@ -50,6 +50,7 @@ export abstract class BaseManagedIdentitySource {
     protected logger: Logger;
     private nodeStorage: NodeStorage;
     private cryptoProvider: CryptoProvider;
+    private networkClient: INetworkModule;
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts:49]: In the end, there is one and only network client. The network client may or may not perform retries. That's the beauty of the decorator pattern.

Why have 2 private variables here - networkClient and networkClientWithRetry?

**Review by bgavrilMS — APPROVED** [review]

File: lib/msal-node/src/config/Configuration.ts [hunk 7]
```diff
@@ -228,20 +240,42 @@ export function buildManagedIdentityConfiguration({
...
                 | http.AgentOptions
                 | https.AgentOptions,
             true // Managed Identity
-        ),
-    };
+        );
+    }
+
+    // wrap the network client with a retry policy if the developer has not disabled the option to do so
+    if (!system?.disableInternalRetries) {
```
Comment (neha-bhargava) [inline lib/msal-node/src/config/Configuration.ts:262]: Is this code common to all? Managed Identity and other calls to ests? If yes, is there somewhere this flag is defaulted to true? Are there tests in node that confirm that only 1 call is made? If not can you add a test to confirm there is no retry in the general flow?

**Review by neha-bhargava — APPROVED** [review]
I have a question about testing `disableInternalRetries` flag for other scenarios. Otherwise looks good. Thanks @Robbie-Microsoft

File: lib/msal-node/src/config/Configuration.ts (hunk 7 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/config/Configuration.ts:262]: This is only for the ManagedIdentity configuration. The flag is defaulted to false.

**Pull request merged**